spec-docs/function-types.md

```diff
@@ -0,0 +1,304 @@
+# Function Types in Kotlin on JVM
+
+## Goals
+
+* Get rid of 23 hardwired physical function classes. One of the problems with them is that they should be effectively duplicated in reflection which means a lot of physical classes in the runtime.
+* Make extension functions coercible to normal functions (with an extra parameter), so that it's possible to do `listOfStrings.map(String::length)`
+* Allow functions with more than 23 parameters, theoretically any number of parameters (in practice 255 on JVM).
+* At the same time, allow to implement Kotlin functions easily from Java: `new Function2() { ... }` and overriding `invoke` only would be the best.
```

Comment:
This means that the situation in which you would only need to implement a single method when passing a lambda from Java to Kotlin is the best. Much better than the situation in which you would  need to override, say, two methods each time (this sounds like nonsense, but still).